app/src/app/api/onchain-weather/route.ts

```diff
@@ -26,13 +33,43 @@ export async function POST(request: NextRequest) {
   if (!params || !params.latitude || !params.longitude)
     return NextResponse.error()
 
-  const result = await requestWeatherOnchain(params)
-  if (!result || !result.requestId) return NextResponse.error()
+  const result = await requestWeatherOnchain({
+    latitude: params.latitude,
+    longitude: params.longitude,
+  })
+  if (!result || !result.requestId || !result.timestamp)
+    return NextResponse.error()
 
   const data = {
     requestId: result.requestId,
     txHash: result.tx.hash,
+    timestamp: result.timestamp,
+  }
+
+  const currentEntries = await kv.lrange<HistoryEntry>('history', 0, -1)
+  if (currentEntries.some((e) => e.txHash === data.txHash)) {
+    return NextResponse.error()
+  }
+  if (currentEntries.length >= 10) {
+    await kv.rpop('history', 1)
+  }
+  const weatherResponse = await fetchCurrentWeather({
+    latitude: params.latitude,
+    longitude: params.longitude,
+  })
+  const weatherCode = getCurrentWeatherCode(weatherResponse)
+  const temperature = getCurrentTemperature(weatherResponse)
+  const temperatureUnit = getCurrentTemperatureUnit(weatherResponse)
+  const historyEntry = {
+    txHash: data.txHash,
+    timestamp: data.timestamp,
+    city: params.city,
+    country: params.country,
+    temperature: `${temperature}`,
+    weatherCode,
+    temperatureUnit,
   }
+  await kv.lpush<HistoryEntry>('history', historyEntry)
```

Comment:
Please extract all history related code to a function. Ideally in a separate module, for example: `lib/history.ts`.